nni/tools/nni_manager_scripts/collect_gpu_info.py

```diff
@@ -0,0 +1,173 @@
+            except Exception as e:
+                errors.add(f'driver_version: {e}')
+
+            try:
+                self.cuda_version = nvmlSystemGetCudaDriverVersion_v2()
+            except Exception as e:
+                errors.add(f'cuda_version: {e}')
+
+        self.gpus = [GpuInfo(index, detail) for index in range(self.gpu_number)]
+
+        procs = []
+        for gpu in self.gpus:
+            procs += gpu._procs
+        self.processes = sorted(procs, key=(lambda proc: proc.pid))
+
+@dataclass(init=False)
+class GpuInfo:
+    index: int
+    model: str | None = None
+    cuda_cores: int | None = None
+    gpu_memory: int | None = None
+    free_gpu_memory: int | None = None
+    gpu_core_utilization: float | None = None
+    gpu_memory_utilization: float | None = None
+    _procs: list[ProcessInfo]
+
+    def __init__(self, index: int, detail: bool):
+        self.index = index
+        self._procs = []
+
+        try:
+            device = nvmlDeviceGetHandleByIndex(self.index)
+        except Exception as e:
+            errors.add(f'device: {e}')
+            return
+
+        if detail:
+            try:
+                self.model = nvmlDeviceGetName(device)
+            except Exception as e:
+                errors.add(f'model: {e}')
+
+            try:
+                self.cuda_cores = nvmlDeviceGetNumGpuCores(device)
+            except Exception as e:
+                errors.add(f'cuda_cores: {e}')
+
+            try:
+                mem = nvmlDeviceGetMemoryInfo(device)
+                self.gpu_memory = mem.total
+                self.free_gpu_memory = mem.free
+            except Exception as e:
+                errors.add(f'gpu_memory: {e}')
+
+        try:
+            util = nvmlDeviceGetUtilizationRates(device)
+            self.gpu_core_utilization = util.gpu / 100
+            self.gpu_memory_utilization = util.memory / 100
+        except Exception as e:
+            errors.add(f'gpu_utilization: {e}')
+
+        try:
+            cprocs = nvmlDeviceGetComputeRunningProcesses_v3(device)
+            gprocs = nvmlDeviceGetGraphicsRunningProcesses_v3(device)
+            self._procs += [ProcessInfo(proc, self.index, 'compute', detail) for proc in cprocs]
+            self._procs += [ProcessInfo(proc, self.index, 'graphics', detail) for proc in gprocs]
+        except Exception as e:
+            errors.add(f'process: {e}')
+
+@dataclass(init=False)
+class ProcessInfo:
+    pid: int
+    name: str | None = None
+    gpu_index: int
+    type: Literal['compute', 'graphics']
+    used_gpu_memory: int | None
+
+    def __init__(self, info: Any, gpu_index: int, type_: Literal['compute', 'graphics'], detail: bool):
+        self.pid = info.pid
+        if detail:
+            self.name = nvmlSystemGetProcessName(self.pid)
+        self.gpu_index = gpu_index
+        self.type = type_
+        self.used_gpu_memory = info.usedGpuMemory
+
+if __name__ == '__main__':
+    main()
```

Comment:
Missing EOL at EOF.

Reply:
I believe my editor adds EOL (and hides EOL) automatically. If it's missing, there will be a red "⏎" line on github web.